draft-west-cookie-incrementalism.md

```diff
@@ -401,85 +404,303 @@ from "have the same registrable domain" (or "is an exact match for") to say
+different scheme.
+
+How we accomplish this is as follows:
+
+First, add the concept of port matching which helps to simplify checking
+if a cookie would match a port value. We can do that by adding a new
+section under 5.1 (this new section depends on the modification to 5.4
+below)
 
 ~~~
-2.   Create a new cookie with name cookie-name, value cookie-value.
-     Set the creation-time and the last-access-time to the current
-     date and time.
+5.1.5 Port matching 
+
+An integer port-matches a given cookie if any of the following
+conditions are true:
+  1. The integer exactly matches the cookie’s port value.
+
+  2. The cookie’s host-only-flag is false.
 ~~~
 
-to:
+We also want to add a sub-algorithm to protect origin cookies from being
+overwritten/shadowed by domain cookies. This’ll be run during the
+creation of new cookies in a later step.
 
 ~~~
-2.   Create a new cookie with name cookie-name, value cookie-value.
-     Set the creation-time and the last-access-time to the current
-     date and time. Set the scheme to request-uri's origin's scheme
-     component.
+5.1.6 New Cookie does not Shadow or Overwrite Origin-only Cookies
+
+This algorithm protects any existing origin-only cookies from being
+overwritten or shadowed by a new cookie ("new-cookie") with the Domain
+attribute and is as follows:
+
+If the new-cookie's host-only-flag is true, then return true.
+
+If the new-cookie’s host-only-flag is false, then return false if the
+cookie store contains at least one cookie ("equiv-cookie") that meets
+all of the following criteria: 
+  1. Equiv-cookie’s name and scheme match new-cookie’s name and scheme.
+
+  2. Equiv-cookie’s host-only-flag is true.
+
+  3. Equiv-cookie’s domain domain-matches the domain of new-cookie. 
+  (Note: this is an asymmetrical comparison) 
+
+Otherwise return true
+
+Notice the asymmetric comparison during the domain-matching. It can be
+helpful to think of domain-matching as: “A domain-matches B when A is
+equivalent to or a sub-domain of B”. Since a cookie with the Domain
+attribute is sent to all hosts that are equivalent to or a sub-domain
+of its Domain’s value, this asymmetric comparison prevents a cookie
+with the Domain attribute from being set when it would be sent to any
+host that also contains an origin-only cookie.
+
+Note: This comparison intentionally ignores the “path” component. The
+intent is to protect a more tightly scoped origin bound cookie across
+the entire origin.
+
+Example: https://landing.site.example sets a cookie `Set-Cookie:
+a=origincookie`. The user then visits https://account.site.example
+which tries to set a cookie `Set-Cookie: a=domaincookie;
+Domain=site.example`. The second cookie would fail to be set because
+the cookie would shadow“a=origincookie”.
+~~~
+
+Next, alter the storage model in Section 5.4 of {{RFC6265bis}} by
+adding“scheme” and “port” to the list of fields the user agent stores
+about each cookie, and setting them. Also, lets refer to the newly
```

Comment:
nit: Word choice/punctuation with "lets"? Not sure if you meant "let's" as in "let us" or "lets" as in "this lets us".